### PR TITLE
add new prediction to bottom of hot flip

### PR DESCRIPTION
**Pull request by jonborchardt**
image

### CODE REVIEW

File: ui/src/components/Hotflip.tsx [hunk 1]
```diff
@@ -34,32 +34,40 @@ const highlightFlippedTokens = (originalInput: string[], flippedInput: string[])
...
+    best_span_str: string;
+    loss: number;
+    passage_question_attention: number[][];
+    passage_tokens: string[];
+    question_tokens: string[];
+    span_end_logits: number[];
+    span_end_probs: number[];
+    span_start_logits: number[];
+    span_start_probs: number[];
+    token_offsets: number[][];
```
Comment (codeviking) [inline ui/src/components/Hotflip.tsx:51]: This looks like it's just another instance of the stuff that's in `types.ts` (the various `Output` subtypes).

Could we just use those types instead?

File: ui/src/components/Hotflip.tsx [hunk 2]
```diff
@@ -34,32 +34,40 @@ const highlightFlippedTokens = (originalInput: string[], flippedInput: string[])
...
 }
 
-export const Hotflip = ({ final, original, new_prediction, context }: HotflipAttackOutput) => {
+export const Hotflip = ({ final, original, outputs, context }: HotflipAttackOutput) => {
+    const finalIndexToShow = 0;
     const [originalString, flippedString] = highlightFlippedTokens(
         original,
-        final[0] // just the first one, huh,... yuck
+        final[finalIndexToShow]
```
Comment (codeviking) [inline ui/src/components/Hotflip.tsx:58]: What if `final` is empty?

File: ui/src/components/Hotflip.tsx [hunk 3]
```diff
@@ -34,32 +34,40 @@ const highlightFlippedTokens = (originalInput: string[], flippedInput: string[])
...
-
-            {new_prediction ? (
-                <p>
-                    <h6>New Prediction:</h6> {new_prediction}
-                </p>
+            <h6>Original Input:</h6> {originalString}
+            <h6>Flipped Input:</h6> {flippedString}
+            {outputs.length && outputs[0].best_span_str ? (
+                <>
+                    <h6>New Prediction:</h6> {outputs[finalIndexToShow].best_span_str}
```
Comment (codeviking) [inline ui/src/components/Hotflip.tsx:68]: I'm confused -- are `outputs` and `final` the same?

Comment (codeviking) [reply ui/src/components/Hotflip.tsx:68]: Also, what if `outputs` is empty?

**Review by codeviking — APPROVED** [review]
LGTM; Seems like we may want to handle the case where the various lists we're pulling things out of are empty.

File: ui/src/components/Hotflip.tsx (hunk 1 above)
Comment (jonborchardt) [reply ui/src/components/Hotflip.tsx:51]: well... you are pointing out a larger issue
i need to fix this up
great find

File: ui/src/components/Hotflip.tsx (hunk 3 above)
Comment (jonborchardt) [reply ui/src/components/Hotflip.tsx:68]: 2 sibling arrays that are in `sync`

File: ui/src/components/Hotflip.tsx (hunk 1 above)
Comment (jonborchardt) [reply ui/src/components/Hotflip.tsx]: please see new code

File: ui/src/components/Hotflip.tsx (hunk 2 above)
Comment (jonborchardt) [reply ui/src/components/Hotflip.tsx]: fixed

File: ui/src/components/InputReduction.tsx [hunk 4]
```diff
@@ -132,17 +132,13 @@ const DefaultOutput = ({
                 return (
                     <>
                         {!formattedOriginal ? (
-                            <p>
-                                <h6>Original Input:</h6> {originalColored}
-                            </p>
+                            <>
```
Comment (jonborchardt) [inline ui/src/components/InputReduction.tsx:135]: while in rome, i fixed this

File: ui/src/demos/reading-comprehension/Predictions.tsx [hunk 5]
```diff
@@ -82,6 +83,14 @@ const OutputByModel = ({
     throw new UnexpectedModelError(model.id);
 };
 
+const BasicAnswer = ({ output }: { output: Prediction }) => {
```
Comment (jonborchardt) [inline ui/src/demos/reading-comprehension/Predictions.tsx:86]: while in rome, i added this

File: ui/src/components/Hotflip.tsx [hunk 6]
```diff
@@ -31,35 +37,30 @@ const highlightFlippedTokens = (originalInput: string[], flippedInput: string[])
...
 }
 
-export const Hotflip = ({ final, original, new_prediction, context }: HotflipAttackOutput) => {
-    const [originalString, flippedString] = highlightFlippedTokens(
-        original,
-        final[0] // just the first one, huh,... yuck
-    );
+interface Props {
+    final?: string[];
+    original: string[];
```
Comment (codeviking) [inline ui/src/components/Hotflip.tsx:48]: I might name these to `originalTokens` and `finalTokens` (if that's what they are) for consistency.

File: ui/src/demos/reading-comprehension/Attacks.tsx [hunk 7]
```diff
@@ -62,7 +63,7 @@ export const Attacks = ({ model, input, target }: Props) => {
                 ) : null}
                 {supportedAttackTypes.has(AttackType.HotFlip) ? (
                     <Collapse.Panel key={AttackType.HotFlip} header="HotFlip">
-                        <Attack<Input, HotflipAttackOutput>
+                        <Attack<Input, HotflipAttackOutput<typeof prediction>>
```
Comment (codeviking) [inline ui/src/demos/reading-comprehension/Attacks.tsx:66]: Shouldn't this just be `Prediction`? There's no way the compiler can infer anything more specific than that, right?

Comment (jonborchardt) [reply ui/src/demos/reading-comprehension/Attacks.tsx:66]: i was surprised it let me do it.
if its cleaner ill convert

**Review by codeviking — APPROVED** [review]
LGTM.

File: ui/src/demos/reading-comprehension/Attacks.tsx (hunk 7 above)
Comment (codeviking) [reply ui/src/demos/reading-comprehension/Attacks.tsx:66]: Yea, I'd just use `Prediction`, as that's what `typeof prediction` will always return...I think?

**Pull request merged**